README.md
# Japanese Date plugin for CraftCMS 3.x

Converts a date string or date object to Japanese format, with the ability to calculate and display the Modern Japan Era name (Reiwa/令和, Heisei/平成, Showa/昭和, Taisho/大正, Meiji/明治).

## Examples

All examples can accept a year as a string or integer (eg. `"2018"` or `2018`), a date string (`"2018-12-25"`), or a date object (eg. `now`).

### Get object with era information
```twig
{{ craft.japaneseDate.era('2018') }} or 
{{ craft.japaneseDate.getEra('2018') }}
```

### Display Era

#### Kanji Name
```twig
{{ craft.japaneseDate.era('2018').kanji_name }} or 
{{ craft.japaneseDate.kanjiEra('2018') }} or 
{{ craft.japaneseDate.getKanjiEra('2018') }}
```

Outputs:

```
平成
```

#### Alphabetical Name
```twig
{{ craft.japaneseDate.era('2018').alpha_name }} or 
{{ craft.japaneseDate.alphaEra('2018') }} or 
{{ craft.japaneseDate.getAlphaEra('2018') }}
```

Outputs:

```
Heisei
```

#### Short Name
```twig
{{ craft.japaneseDate.era('2018').short_name }} or 
{{ craft.japaneseDate.shortEra('2018') }} or 
{{ craft.japaneseDate.getShortEra('2018') }}
```

Outputs:

```
H
```

### Display Year

```twig
{{ craft.japaneseDate.year('2018') }} or {{ craft.japaneseDate.getYear('2018') }}
```

Same as (alias of):

```twig
{{ craft.japaneseDate.era('2018') }}
```

#### Year in Kanji
```twig
{{ craft.japaneseDate.kanjiYear('2018') }} or 
{{ craft.japaneseDate.getKanjiYear('2018') }}
```

Outputs:

```
平成30年
```

Use `true` as a second parameter to use 元 for Year 1 of an era:

```twig
{{ craft.japaneseDate.kanjiYear('1989', true) }} or 
{{ craft.japaneseDate.getKanjiYear('1989', true) }}
```

Outputs:

```
平成元年
```

#### Year in Short Format
```twig
{{ craft.japaneseDate.shortYear('2018') }} or 
{{ craft.japaneseDate.getShortYear('2018') }}
```

Outputs:

```
H30
```

### Full Date
```twig
{{ craft.japaneseDate.date('2018-12-25') }} or 
{{ craft.japaneseDate.getDate('2018-12-25') }}
```

Outputs:

```
平成30年12月25日
```

Use `true` as a second parameter to use 元 for Year 1 of an era:

```twig
{{ craft.japaneseDate.date('1989-12-25', true) }} or 
{{ craft.japaneseDate.getDate('1989-12-25', true) }}
```

Outputs:

```
平成元年12月25日
```


CHANGELOG.md
# Release Notes for Japanese Date

## 1.0.0 - 2020-03-11

- Initial commit

src/variables/JapaneseDateVariable.php
<?php
/**
 * Created by PhpStorm.
 * User: nicolasbottari
 * Date: 2020/03/04
 * Time: 17:18
 */

namespace zenbustudio\japanesedate\variables;

use DateTime;
use zenbustudio\japanesedate\JapaneseDate;
use zenbustudio\japanesedate\models\YearModel;

class JapaneseDateVariable {

	/**
	 * @param $date
	 *
	 * @return \zenbustudio\japanesedate\models\YearModel
	 */
	public function getEra($date): YearModel
	{
		if($date instanceof DateTime)
		{
			$dateObject = $date;
		}
		elseif(is_numeric($date) && strlen($date) == 4)
		{
			$dateObject = new DateTime($date.'-01-01');
		}
		else
		{
			$dateObject = new DateTime($date);
		}

		$year = $dateObject->format('Y');
		$era = JapaneseDate::getInstance()->JapaneseDate->findEra($year);
		return $era;
	}

	/**
	 * @param $date
	 *
	 * @return string
	 */
	public function getAlphaEra($date): string
	{
		return $this->getEra($date)->alpha_name;
	}

	/**
	 * @param $date
	 *
	 * @return string
	 */
	public function getKanjiEra($date): string
	{
		return $this->getEra($date)->kanji_name;
	}

	/**
	 * @param $date
	 *
	 * @return string
	 */
	public function getShortEra($date): string
	{
		return $this->getEra($date)->short_name;
	}

	/**
	 * @param $date
	 *
	 * @return \zenbustudio\japanesedate\models\YearModel
	 */
	public function getYear($date): YearModel
	{
		return $this->getEra($date);
	}

	/**
	 * @param $date
	 *
	 * @return string
	 */
	public function getShortYear($date): string
	{
		$yearModel = $this->getEra($date);
		return $yearModel->short_name . $yearModel->years_in_era;
	}

	/**
	 * @param      $date
	 * @param bool $use_gannen
	 *
	 * @return string
	 */
	public function getKanjiYear($date, bool $use_gannen = false): string
	{
		$yearModel = $this->getEra($date);

		if($use_gannen && $yearModel->years_in_era == 1)
		{
			return $yearModel->kanji_name . $yearModel::KANJI_FIRST_ERA_YEAR . $yearModel::KANJI_YEAR;
		}

		return $yearModel->kanji_name . $yearModel->years_in_era . $yearModel::KANJI_YEAR;
	}

	/**
	 * @param      $date
	 * @param bool $use_gannen
	 *
	 * @return string
	 */
	public function getDate($date, bool $use_gannen = false): string
	{
		if($date instanceof DateTime)
		{
			$dateObject = $date;
		}
		elseif(is_numeric($date) && strlen($date) == 4)
		{
			$dateObject = new DateTime($date.'-01-01');
		}
		else
		{
			$dateObject = new DateTime($date);
		}

		$month = $dateObject->format('m');
		$day = $dateObject->format('d');

		return $this->getKanjiYear($date, $use_gannen) . $month . '月' . $day . '日';
	}
}

src/services/JapaneseDateService.php
<?php
/**
 * Created by PhpStorm.
 * User: nicolasbottari
 * Date: 2020/03/04
 * Time: 23:02
 */

namespace zenbustudio\japanesedate\services;

use Craft;
use zenbustudio\japanesedate\models\YearModel;

class JapaneseDateService {

	public $start_years = [
		1868 => 'meiji',
		1912 => 'taisho',
		1926 => 'showa',
		1989 => 'heisei',
		2019 => 'reiwa',
	];

	public $era_data = [
		'meiji'  => [
			'alpha_name' => 'Meiji',
			'kanji' => '明治',
			'short' => 'M',
		],
		'taisho' => [
			'alpha_name' => 'Taisho',
			'kanji' => '大正',
			'short' => 'T',
		],
		'showa'  => [
			'alpha_name' => 'Showa',
			'kanji' => '昭和',
			'short' => 'S',
		],
		'heisei' => [
			'alpha_name' => 'Heisei',
			'kanji' => '平成',
			'short' => 'H',
		],
		'reiwa'  => [
			'alpha_name' => 'Reiwa',
			'kanji' => '令和',
			'short' => 'R',
		],
	];

	/**
	 * @param string $year
	 *
	 * @return \zenbustudio\japanesedate\models\YearModel
	 */
	public function findEra(string $year): YearModel
	{
		$found_era_name = null;

		foreach($this->start_years as $start_year => $era_name)
		{
			if($year >= $start_year)
			{
				$found_era_name = $era_name;
			}
		}

		$yearModel = new YearModel;

		if(isset($this->era_data[$found_era_name]))
		{
			$start_year = array_search($found_era_name, $this->start_years);
			$yearModel->kanji_name = $this->era_data[$found_era_name]['kanji'];
			$yearModel->alpha_name = $this->era_data[$found_era_name]['alpha_name'];
			$yearModel->short_name = $this->era_data[$found_era_name]['short'];
			$yearModel->start_year = $start_year;
			$yearModel->years_in_era = $year - $start_year + 1;
			$yearModel->is_first_year = $this->isFirstYear($year);
		}

		return $yearModel;
	}

	/**
	 * @param string $year
	 *
	 * @return bool
	 */
	public function isFirstYear(string $year): bool
	{
		foreach($this->start_years as $start_year => $era_name)
		{
			if($year == $start_year)
			{
				return true;
			}
		}

		return false;
	}
}